presto-bigquery/src/main/java/com/facebook/presto/bigquery/BigQueryResultPageSource.java

```diff
@@ -0,0 +1,342 @@
+            type.writeSlice(output, Decimals.encodeScaledValue(bdValue, NUMERIC_DATA_TYPE_SCALE));
+        }
+        else if (type instanceof VarbinaryType) {
+            if (value instanceof ByteBuffer) {
+                type.writeSlice(output, Slices.wrappedBuffer((ByteBuffer) value));
+            }
+            else {
+                output.appendNull();
+            }
+        }
+        else {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unhandled type for Slice: " + type.getTypeSignature());
+        }
+    }
+
+    private void writeBlock(BlockBuilder output, Type type, Object value)
+    {
+        if (type instanceof ArrayType && value instanceof List<?>) {
+            BlockBuilder builder = output.beginBlockEntry();
+
+            for (Object element : (List<?>) value) {
+                appendTo(type.getTypeParameters().get(0), element, builder);
+            }
+
+            output.closeEntry();
+            return;
+        }
+        if (type instanceof RowType && value instanceof GenericRecord) {
+            GenericRecord record = (GenericRecord) value;
+            BlockBuilder builder = output.beginBlockEntry();
+
+            List<String> fieldNames = new ArrayList<>();
+            for (int i = 0; i < type.getTypeSignature().getParameters().size(); i++) {
+                TypeSignatureParameter parameter = type.getTypeSignature().getParameters().get(i);
+                fieldNames.add(parameter.getNamedTypeSignature().getName().orElse("field" + i));
+            }
+            checkState(fieldNames.size() == type.getTypeParameters().size(), "fieldName doesn't match with type size : %s", type);
+            for (int index = 0; index < type.getTypeParameters().size(); index++) {
+                appendTo(type.getTypeParameters().get(index), record.get(fieldNames.get(index)), builder);
+            }
+            output.closeEntry();
+            return;
+        }
+        throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unhandled type for Block: " + type.getTypeSignature());
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        bigQueryStorageClient.close();
+        closed = true;
+    }
+
+    Iterable<GenericRecord> parse(Storage.ReadRowsResponse response)
+    {
+        byte[] buffer = response.getAvroRows().getSerializedBinaryRows().toByteArray();
+        readBytes.addAndGet(buffer.length);
+        log.debug("Read %d bytes (total %d) from %s", buffer.length, readBytes.get(), split.getStreamName());
+        Schema avroSchema = new Schema.Parser().parse(split.getAvroSchema());
+        return () -> new AvroBinaryIterator(avroSchema, buffer);
+    }
+
+    Stream<GenericRecord> toRecords(Storage.ReadRowsResponse response)
+    {
+        byte[] buffer = response.getAvroRows().getSerializedBinaryRows().toByteArray();
```

Comment:
just curious, why have to use Avro?

Reply:
It was the default at the time from the spark and prestosql connectors, also avro supports disjunction push down while arrow does not